Renderer `rebuildCells` rework

qwerasd205 opened the pull request:
This was a long time coming and I went ahead and did it while fixing a bug with background colors behind ligatures.

This is a significant rework of the renderer cell rebuilding, removing a lot of duplicated work from inside of `updateCell` and making things generally slightly more sane.

Remember this performance graph from the last time I improved the shaper performance?

<img width="951" alt="image" src="https://github.com/user-attachments/assets/292c726c-efcf-45e9-8257-f9809a94b8cc">

:)

> [!NOTE]
> I am unable to test the preedit cell logic for the OpenGL renderer because I'm on macOS.
> Someone else will need to do that before this can be merged.

qwerasd205 committed:
renderer, shaper: don't use null cells, handle bg and decorations separately

Significant rework that also removes a lot of unnecessarily duplicated
work while rebuilding cells in both renderers. Fixes multiple issues
with decorations and bg colors on wide chars and ligatures, while
reducing the amount of special case handling required.

qwerasd205 committed:
coretext: don't emit 0 codepoints for special fonts

mitchellh reviewed (changes requested):
I haven't tried to _understand_ it yet, but here is some superficial feedback since I think you're taking a look at it anyways. Thanks for this!

qwerasd205 committed:
test(coretext): add test for high plane padding sprite font behavior

qwerasd205 committed:
renderer: cleanup, reduce nesting, more comments

qwerasd205 committed:
coretext: fix tests to account for removal of null cells

qwerasd205 committed:
harfbuzz: fix tests to account for removal of null cells

mitchellh committed:
font/noop: conform to new run struct type

qwerasd205 committed:
font/shaper: explicitly skip invisible cells while shaping

Fixes a bug caused by the renderer logic assuming this behavior and not
advancing the run iterator when skipping foreground elements in cells
with the invisible flag set.
